src/pymor/algorithms/projection.py

```diff
@@ -242,6 +243,23 @@ def action_BlockOperatorBase(self, op):
         else:
             return np.sum(projected_ops)
 
+    @match_class(BilinearFunctional)
+    def action_BilinearFunctional(self, op):
+        if op.operator.source == op.operator.range:
+            return BilinearFunctional(project(op.operator, self.source_basis, self.source_basis), name=op.name)
+        else:
+            return BilinearFunctional(project(op.operator, self.range_basis, self.source_basis), name=op.name)
+
+    @match_class(BilinearProductFunctional)
+    def action_BilinearProductFunctional(self, op):
+        op1 = project(op.operators[0], self.range_basis, self.source_basis)
```

Comment:
That can't be right: `range_basis` would be in the `range` of the operator to project (the `BilinearProductFunctional`), but that maps to `NumpyVectorSpace(1)`. If one wanted to preserve the structure of the operator during projection, one would probaby need to use `estimate_image` to build a common basis for the range of both operators and than use that as `range_basis`. Probably not worth the effort, IMHO, unless one has a specific application where one needs to preserve structure.

One the other hand, one should check if `project` works properly with the generic rule for `BilinearFunctional` when both `A` and `B` are affinely decomposed and a `product` is given. (I would hope so, but it might be good to check again.)

Reply:
Right, I was somehow thinking about projecting the inner operators with their own range bases and then multiplying them afterwards. But that's complete bollocks of course!